core/src/main/java/org/apache/calcite/rel/rules/ProjectTableScanRule.java

```diff
@@ -103,30 +106,47 @@ protected static boolean test(TableScan scan) {
   protected void apply(RelOptRuleCall call, Project project, TableScan scan) {
     final RelOptTable table = scan.getTable();
     assert table.unwrap(ProjectableFilterableTable.class) != null;
-    if (!project.isMapping()) {
-      return;
-    }
-    final Mappings.TargetMapping mapping = Project.getPartialMapping(
-            project.getInput().getRowType().getFieldCount(),
-            project.getProjects());
 
-    final ImmutableIntList projects;
-    final ImmutableList<RexNode> filters;
+    final List<Integer> selectedColumns = new ArrayList<>();
+    project.getProjects().forEach(proj -> {
+      proj.accept(new RexVisitorImpl<Void>(true) {
+        public Void visitInputRef(RexInputRef inputRef) {
+          if (!selectedColumns.contains(inputRef.getIndex())) {
+            selectedColumns.add(inputRef.getIndex());
+          }
+          return null;
+        }
+      });
+    });
+
+    final List<RexNode> filtersPushDown;
+    final List<Integer> projectsPushDown;
     if (scan instanceof Bindables.BindableTableScan) {
       final Bindables.BindableTableScan bindableScan =
           (Bindables.BindableTableScan) scan;
-      filters = bindableScan.filters;
-      projects = bindableScan.projects;
+      filtersPushDown = bindableScan.filters;
+      projectsPushDown = selectedColumns.stream()
+          .map(col -> bindableScan.projects.get(col)).collect(Collectors.toList());
     } else {
-      filters = ImmutableList.of();
-      projects = scan.identity();
+      filtersPushDown = ImmutableList.of();
+      projectsPushDown = selectedColumns;
     }
+    Bindables.BindableTableScan newScan = Bindables.BindableTableScan.create(
+        scan.getCluster(), scan.getTable(), filtersPushDown, projectsPushDown);
+    final List<RexNode> newProjectRexNodes = new RexShuttle() {
+      @Override public RexNode visitInputRef(RexInputRef inputRef) {
+        int newIdx = selectedColumns.indexOf(inputRef.getIndex());
+        assert newIdx != -1;
+        return new RexInputRef(newIdx, inputRef.getType());
+      }
+    }.apply(project.getProjects());
```

Comment:
Can we use mapping to mutate the projects?

Reply:
Thanks ~ that would be better ~